### PR TITLE
fa-ir localization

### PR DESCRIPTION
**Pull request by HKalbasi**
Following https://github.com/MycroftAI/mycroft-core/pull/2778 I translated this library to Farsi.
For CLA, my email is hamidrezakalbasi at protonmail.com and I'm working as an individual.

### CODE REVIEW

**Comment by devs-mycroft** [comment]
Hello, @HKalbasi, thank you for helping with the Mycroft project! We welcome everyone
into the community and greatly appreciate your help as we work to build an AI
for Everyone.

To protect yourself, the project, and users of Mycroft technologies we require
a Contributor Licensing Agreement (CLA) before accepting any code
contribution. This agreement makes it crystal clear that along with your
code you are offering a license to use it within the confines of this project.
You retain ownership of the code, this is just a license.

Please visit https://mycroft.ai/cla to initiate this one-time signing. Thank
you!

**Comment by ChanceNCounter** [comment]
Hi, and thanks for helping out! I often feel compelled to say up front, I don't work for MycroftAI. I'm a community collaborator on this repository, and there's about a 1/4 likelihood I'll be the one who ends up reviewing your code. Pleased to meet you!

Reviewing unfamiliar languages is challenging for exactly the same reason we need and appreciate help from contributors around the world, and who speak a variety of languages: we don't speak the language we're reviewing 😛

This is also the first time I've been presented with an unfamiliar alphabet. It's always been coming, and Polish was a step in that direction, but here's the real deal!

For that reason, it would be cool if you could get a couple other Farsi speakers to verify the grammatical correctness and accuracy of the phrases in your test files. If you can't find anybody, that's no problem - wouldn't be our first leap of faith - but it's a nice security blanket. Also helps address the extremely unlikely possibility that somebody submits a file full of vulgarity 😉 

I wanna stress, that's not a dealbreaker, it's just something I'd appreciate since none of us can speak or read Farsi.

**Comment by rimthekid** [comment]
hey I just read test codes and was thinking it's better if you use Persian calendar as well

**Comment by ChanceNCounter** [comment]
@rimthekid that's probably a MUCH bigger undertaking

**Comment by rimthekid** [comment]
I don't see any other problems though

File: lingua_franca/res/text/fa-ir/date_time.json
Comment (ahangarha) [inline lingua_franca/res/text/fa-ir/date_time.json:139]: مارس
آوریل

**Review by ahangarha — CHANGES_REQUESTED** [review]

File: lingua_franca/lang/common_data_fa.py
Comment (senaps) [inline lingua_franca/lang/common_data_fa.py:193]: I may have been away from school for too long, but I'm pretty sure that we wouldn't have read `15` as ‍`پونزده` and it was `پانزده` same goes for 16, 17 and 18 representations too.

**Review by reza-keshavarz — COMMENTED** [review]
عبارت «سیم» یه ذره مبهمه. «سی ام» اگر نوشته بشه یه ذره بهتر می‌تونه باشه.

File: lingua_franca/lang/common_data_fa.py
Comment (HKalbasi) [reply lingua_franca/lang/common_data_fa.py:193]: `پونزده` is used in conversation, but `پانزده` is formal. Which one should we use @ChanceNCounter ?

File: lingua_franca/lang/common_data_fa.py
Comment (senaps) [inline lingua_franca/lang/common_data_fa.py]: these guys shouldn't have been translated in this way, that's an informal/casual way of saying the words, so, although we might encounter the words and numbers in this way more often, the actual representation is `سی‌ام` instead of `سیم‍`. but, there is another `سی ام` thing going on too. (hassles of writing and speaking Persian I guess?)
and, we actually have a translation for `wire` for the word `سیم` too if that's not complicated enough.

so, there is a compromise to be made, or a solution should be provided since there are sometimes up to 3-4 ways of writing the same word and all are true.

Comment (HKalbasi) [reply lingua_franca/lang/common_data_fa.py]: In parsing I handle (or tried to handle) all cases but in formatting I think we have to choose one of the styles.

**Review by rimthekid — APPROVED** [review]

File: lingua_franca/lang/common_data_fa.py
Comment (JarbasAl) [reply lingua_franca/lang/common_data_fa.py]: look at the catalan PR, i added a decorator for a similar use case, theres 4 different ways to say the time in catalan, this is similar

i suppose this could be used to let user select which variant to use, will require some reworking of the internals so maybe leave it for a follow up PR

Comment (ChanceNCounter) [reply lingua_franca/lang/common_data_fa.py]: We (read: @JarbasAl) recently introduced a function decorator to help in situations like this, at `internal.lookup_variant()`, but I couldn't possibly say if it would help in this *specific* situation.

Comment (HKalbasi) [reply lingua_franca/lang/common_data_fa.py]: So I should add a parameter `variant` to format functions with two possible value `formal` (for `پونزدهم`) and `conversational` (for `پانزدهم`) ?
In case (`سیم` vs `سی ام`) I think it does not matter so much in formatting, The pronunciation of both of them are equal and we can choose one of them (I prefer `سی ام`). But `پونزدهم` is important.

File: lingua_franca/lang/common_data_fa.py
Comment (ChanceNCounter) [reply lingua_franca/lang/common_data_fa.py:61]: This has been the subject of discussion in some other languages. @JarbasAl what do you think? My native language doesn't really have that distinction

Comment (JarbasAl) [reply lingua_franca/lang/common_data_fa.py:61]: its tricky, its for voice usage so should be conversational, but in general mycroft should be somewhat formal. for now i think only a native speaker can decide what makes for a better default, but i tend towards formal

using the variant decorator can allow users to change this behavior later, by default we should be the most respectful possible

File: lingua_franca/lang/common_data_fa.py
Comment (JarbasAl) [reply lingua_franca/lang/common_data_fa.py]: yep exactly, in whatever methods you think it makes sense you can use that decorator. catalan language is the only usage in the wild so far and is a good template

but keep in mind these are not yet used by mycroft-core, a follow up PR will be needed to actually select those variants depending on user config. But the first step is getting the functionality here ;)

**Comment by JarbasAl** [comment]
really cool to see more languages being added! thank you for your work and keep the PRs coming

File: test/test_parse_fa.py
Comment (ahangarha) [inline test/test_parse_fa.py:71]: شانزده هزار و صد و پونزده؟  
بالاخره رسمی یا عامیانه؟

One part is formal and the other is informal. It is intentional?

File: lingua_franca/lang/common_data_fa.py
Comment (ahangarha) [reply lingua_franca/lang/common_data_fa.py:61]: I would love to have different flavors but I believe that the formal language should be the default.

File: test/test_parse_fa.py
Comment (HKalbasi) [reply test/test_parse_fa.py:71]: This is a test for parsing. User input can contain both forms And we have to handle both. This test checks that parser implemented this correctly.

File: lingua_franca/lang/format_fa.py
Comment (HKalbasi) [inline lingua_franca/lang/format_fa.py:50]: @JarbasAl @ChanceNCounter I tried to handle variant formats with a local decorator, but strangely, kwargs is empty always! (so tests are failing). I did not understand where the problem came from. Can you take a look?

File: lingua_franca/lang/format_fa.py
Comment (JarbasAl) [inline lingua_franca/lang/format_fa.py:57]: im not sure i understand what you are trying to do, this should be

```
@lookup_variant({
           "default": NumberVariantFA.CONVERSATIONAL,
	       "conversational": NumberVariantFA.CONVERSATIONAL,
	        "formal": NumberVariantFA.FORMAL
        })
def nice_number_fa(number, speech=True, denominators=range(1, 21), variant=None):
```

File: lingua_franca/lang/format_fa.py
Comment (JarbasAl) [reply lingua_franca/lang/format_fa.py:50]: see my other comment, you are not using the lookup_variant decorator like i intended it to be used, the logic to handle the variant should be inside the the nice_number method

```python
if variant == NumberVariantFA.FORMAL:
    # changes required for this variant
elif variant == NumberVariantFA.CONVERSATIONAL:
    # changes required for this variant
```

File: lingua_franca/lang/format_fa.py
Comment (HKalbasi) [reply lingua_franca/lang/format_fa.py:57]: I tried not to have duplicate code. I wanted to create the format in the same decorator by replacing the numbers in result and not repeat this in all the outputs of all the functions. That's why I made another decorator.

Comment (JarbasAl) [reply lingua_franca/lang/format_fa.py:57]: oh, i didnt realize this was used in more methods, let me look into it :+1:

you are right code should not be duplicated without a good reason

File: lingua_franca/lang/common_data_fa.py
Comment (JarbasAl) [inline lingua_franca/lang/common_data_fa.py:45]: can all these be made UPPERCASE so it's consistent across all languages?

File: lingua_franca/lang/format_fa.py
Comment (JarbasAl) [reply lingua_franca/lang/format_fa.py:57]: @HKalbasi can you point me to a couple unittests where i cant test this?

just add new unittest section for testing variants, like [here](https://github.com/Softcatala/lingua-franca/blob/b2cf1919a819bb00b895c1f51f926bc0923fc4a5/test/test_format_ca.py#L343)

that way i can just run those individual tests and verify things work like they are supposed to, i only need 1 test case that should change depending on variant

Comment (HKalbasi) [reply lingua_franca/lang/format_fa.py:57]: I added a test in [this commit](https://github.com/MycroftAI/lingua-franca/pull/171/commits/54ab5b45427a9dfd99fdb75de0227bf76219fbfe)

File: lingua_franca/lang/common_data_fa.py
Comment (HKalbasi) [reply lingua_franca/lang/common_data_fa.py:45]: Please specify all the names that have problems so I can change all of them at the same time.

Comment (JarbasAl) [reply lingua_franca/lang/common_data_fa.py:45]: everything in common_data_XXX essentially, so all dicts and lists in that file

Comment (HKalbasi) [reply lingua_franca/lang/common_data_fa.py:45]: Done

File: lingua_franca/lang/format_fa.py
Comment (HKalbasi) [reply lingua_franca/lang/format_fa.py:57]: @JarbasAI Do you solve this? What should I do?

Comment (HKalbasi) [reply lingua_franca/lang/format_fa.py:64]: I fixed it

**Comment by HKalbasi** [comment]
I handled variants. Any thing else before merge @ChanceNCounter @JarbasAI ?

**Comment by ChanceNCounter** [comment]
I'll defer to @JarbasAl re: the correct implementation of variants. Also, have you added some tests for the variants? I suppose that would confirm that they're implemented correctly regardless =P

If you've got at least 4-5 (ideally 5-10) tests for the variants, you're probably in pretty good shape. It'll need another round of review, but you've definitely been hitting the right buttons as you go, and the consensus among those other Farsi speakers seemed to be that you've done the first pass pretty well.

Remember, when adding a new language to Lingua Franca, that it's never finished! You will discover that your number extractors can only count to a certain number, or your datetime extractors are off by a day, or your formatters mess up on decimals, or *who knows*? That's what the language-specific issue labels are for. Everything will be 200 OK

**Comment by HKalbasi** [comment]
> If you've got at least 4-5 (ideally 5-10) tests for the variants, you're probably in pretty good shape

There is already some test for variant. I added them in 54ab5b4 

> It'll need another round of review, but you've definitely been hitting the right buttons as you go, and the consensus among those other Farsi speakers seemed to be that you've done the first pass pretty well.

I think now everyone should be satisfied

> Remember, when adding a new language to Lingua Franca, that it's never finished!

That's ok! But for the first merge, is there anything else I need to do?

**Comment by ChanceNCounter** [comment]
It might be a good idea to try this branch against [this](https://github.com/MycroftAI/mycroft-core/pull/2772) branch of Mycroft-core (which is Mycroft's pending upgrade to the refactored LF.)

I wouldn't usually worry about a direct comparison with downstream, but I'm wondering how problematic it will (or won't) be that `normalize.json` is not fully translated.

**Comment by ChanceNCounter** [comment]
Still waiting for a CLA label on this one. Idk if that's because the CLA hasn't been completed or because the robot hasn't updated the PR.

**Comment by HKalbasi** [comment]
It doesn't send email to me. I tried many times. My first name is `Hamidreza`, my last name is `Kalbasi` and my email is `hamidrezakalbasi at protonmail.com`.

**Comment by krisgesling** [comment]
Hey not sure what happened but I've just sent you a CLA request.

I'm going to change up the way we do the CLA signings as this is not the first (or second, or third) time this has happened...

**Comment by HKalbasi** [comment]
I signed CLA some days ago but label is still `CLA:needed`

**Comment by krisgesling** [comment]
> I signed CLA some days ago but label is still `CLA:needed`

CLA received and you've been added to our [list of excellent humans](https://github.com/mycroftAI/contributors). Unfortunately we have to manually review signings to make sure they've been completed correctly so it can take a few days particularly over the weekend.

Thanks again for your contributions though!

**Comment by devs-mycroft** [comment]
Hello, @HKalbasi, thank you for helping with the Mycroft project! We welcome everyone
into the community and greatly appreciate your help as we work to build an AI
for Everyone.

To protect yourself, the project, and users of Mycroft technologies we require
a Contributor Licensing Agreement (CLA) before accepting any code
contribution. This agreement makes it crystal clear that along with your
code you are offering a license to use it within the confines of this project.
You retain ownership of the code, this is just a license.

Please visit https://mycroft.ai/cla to initiate this one-time signing. Thank
you!

**Comment by krisgesling** [comment]
Weird... please ignore our friendly but misguided bot...

Will have to look into why it's misbehaving

**Comment by ChanceNCounter** [comment]
*Right*. At last we're getting to bump/merge/bump/merge fun time.

@HKalbasi are you comfortable squashing this manually? I usually `git checkout -b` a backup of the feature branch before I start =P

If not, we'll use GitHub's feature to do it, but we can add version bumps if the branch is squashed already.

I used to do it myself, but I've had trouble lately pushing a rebase without clobbering authorship in GitHub's records. It all stays correct in `git log`, but not on GitHub. Hence, I'm not too comfortable force-pushing other people's branches on GH anymore.

**Comment by HKalbasi** [comment]
I don't know how to squash without github exactly, and I'm ok if authorship being messed. So take it easy and do whatever needed.

**Comment by ChanceNCounter** [comment]
Will do.

At the moment, I'm leaning - gotta confer with others - toward releasing the current `master` as 0.4.0, then immediately merging this and tagging it 0.4.1. Assuming everything is clean at that point, we'll go straight to 0.4.1 on pypi; if it turns out this needs follow-up, we can publish 0.4.0 in the meantime.

This has the happy side effect of letting us maintain our release naming pattern: if a release contains a new language, assuming it's the only new language, we like to name the release "Hello World" in that language. 0.4.0 was already named for Catalan, so moving Farsi to 0.4.1 gets Farsi its own named tag =)

What's the Farsi equivalent of Hello World?

**Comment by HKalbasi** [comment]
=)))
That is "سلام دنیا" or "salaam donya"

File: lingua_franca/lang/format_fa.py
Comment (JarbasAl) [inline lingua_franca/lang/format_fa.py:39]: remove the debug prints

File: lingua_franca/lang/format_fa.py
Comment (JarbasAl) [inline lingua_franca/lang/format_fa.py:47]: can this be renamed `_handle_number_variant`? makes the method name explicit and uses snake_case for consistency

File: lingua_franca/lang/format_fa.py
Comment (JarbasAl) [inline lingua_franca/lang/format_fa.py:57]: remove the debug print

**Comment by ChanceNCounter** [comment]
Reckon I can hit those after the rebase.

**Comment by ChanceNCounter** [comment]
Deep breath, here comes pain or gain...

**Comment by ChanceNCounter** [comment]
...don't mind my github extension closing the PR, it'll come back with those fixups

**Comment by ChanceNCounter** [comment]
The warnings from pytest were because a default full language code was not specified in `internal._DEFAULT_FULL_LANG_CODES`, fixing that with the other little things

**Pull request closed without merging**
